### ISSUE TITLE
Miscompilation of Bevy (and some wgpu) apps resulting in segfault on macOS.

### ISSUE THREAD

**Issue by inodentry** [rust-lang/rust#117902]
Recent rustc seems to miscompile something somewhere in the dependency tree of Bevy, causing apps to crash with a segfault.

The issue seems to only manifest when optimizations are enabled and codegen units are > 1. The default cargo release configuration triggers it. Unoptimized builds are fine. Setting `codegen-units = 1` with `opt-level = 3` also seems to be fine.

I am on macOS 14.1 Sonoma, on Apple M1 Pro hardware.

### Code

You can reproduce with any Bevy application. It is a huge code base, I know, I'm sorry I can't find a more minimal example.

Clone the Bevy repo: https://github.com/bevyengine/bevy

all these revs have the issue:

- commit bf4f4e42da3da07b0e84a4d6e40077f7398aea8b (i had that one laying around)
- commit 749f3d74305d35b4a453a80e45b22f08d5189a48 (current latest)
- tag v0.12.0
- tag v0.11.3

Run some example in release mode:
```
cargo run --example bevymark --release
```

### Version it worked on

The previous 1.74.0-beta.7 did not have the issue. I am not sure which nightly introduced the regression.

### Version with regression

I first noticed the issue when I installed the 2023-11-11 nightly, though it might also be present in earlier nightlies. I just installed 1.75.0-beta.1 and updated my nightly, and can confirm the issue is present there, too. 

```
rustc 1.75.0-beta.1 (782883f60 2023-11-12)
binary: rustc
commit-hash: 782883f609713fe9617ba64d90086742ec62d374
commit-date: 2023-11-12
host: aarch64-apple-darwin
release: 1.75.0-beta.1
LLVM version: 17.0.4
```

```
rustc 1.76.0-nightly (ba7c7a301 2023-11-13)
binary: rustc
commit-hash: ba7c7a301984967c8c13adb580ef9b86ba706a83
commit-date: 2023-11-13
host: aarch64-apple-darwin
release: 1.76.0-nightly
LLVM version: 17.0.4
```

**Comment by Nilstrieb** [comment]
I cloned the repo and ran the command you provided, but got errors:
```
error[E0463]: can't find crate for `bevy_reflect_derive`
 --> crates/bevy_reflect/src/type_uuid.rs:1:9
  |
1 | pub use bevy_reflect_derive::TypeUuid;
  |         ^^^^^^^^^^^^^^^^^^^ can't find crate
```
with both stable and nightly on bevy commit `749f3d74305d35b4a453a80e45b22f08d5189a48`
did I do something wrong? what's your bevy commit?

**Comment by inodentry** [comment]
@Nilstrieb I specifically tried:
 - commit `bf4f4e42da3da07b0e84a4d6e40077f7398aea8b` (i had that one laying around)
 - commit `749f3d74305d35b4a453a80e45b22f08d5189a48` (current latest, same as you)
 - tag `v0.12.0`
 - tag `v0.11.3`
 
Yes, the previous 2 releases of Bevy (at least) are affected, too.

I don't get your error, everything compiles fine on my end. 🤔 

Have you tried `cargo clean` followed by `cargo update`, to ensure you have a clean target directory and cargo isn't getting confused by outdated deps?

**Comment by Nilstrieb** [comment]
uhhh, I have run `cargo clean` and both `cargo update`.... wtf cargo?? I tried it on another computer and it compiles. let's dismiss that as a me issue. I can't try it on this other computer for unrelated reasons though so I can't try reproducing it on linux as well to rule out macos-specific issues.

**Comment by Nilstrieb** [comment]
It would be useful if you could use `cargo-bisect-rustc` to bisect the regression to a single PR, which would make it easier to figure out what's happening and whether it really is a miscompilation or just UB that's now going wrong in bevy's dep tree.

**Comment by Nilstrieb** [comment]
I was not able to reproduce this on Windows. Marking as O-MacOS, but I suspect that the problem is platform-independent and just only manifests in platform-specific bevy code.

**Comment by inodentry** [comment]
OK. TIL about `cargo-bisect-rustc`. Installing it and reading the documentation now. Will try it in a moment.

**Comment by inodentry** [comment]
Bisected down to:
Regression in nightly-2023-10-19

It is now checking the individual commits. Will post results when ready. :)

**Comment by inodentry** [comment]
Bisected down to:
Regression in nightly-2023-10-19
Regression in 5d5edf0248d967baa6ac5cbea09b91c7c9947942

**Comment by Nilstrieb** [comment]
oh well, that is sadly not a very useful bisection. it enabled more inlining, causing more optimizations to be applied overall. @saethlin anyways cause you may be interested in this, but I can't imagine that your PR is at fault
can you try enabling LTO, which should get the same behavior on earlier versions and see whether that crashes as well and then bisect that?
Additionally, it would be useful to use a debugger to figure out where the segfault occurred and maybe why it did. It may be possible to figure out what got compiled weirdly, which could lead us to be able to figure out whether it's UB in the code.

**Comment by saethlin** [comment]
That makes me wonder if Bevy is doing this thing that Embassy is (was?) doing: https://github.com/rust-lang/rust/issues/117047

**Comment by inodentry** [comment]
@Nilstrieb Just tested it. LTO does *not* cause a segfault in earlier versions, and *does* cause a segfault in new versions. It does not seem to affect the results.

Also, like I said, setting `codegen-units = 1` in the new rustc does *not* result in segfault. Even though it should be enabling the most inlining optimizations. It is quite interesting.

The issue occurs only with optimized builds with >1 codegen units, regardless of LTO.

**Comment by inodentry** [comment]
Ran in lldb (im not sure how to make it demangle / prettify symbol names):

```
* thread #1, name = 'main', queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x4b8c64f1019f8cf8)
    frame #0: 0x00000001013c70dc bevyregress`wgpu_types::TextureFormat::required_features::h5d814caf5b611efe + 20
bevyregress`wgpu_types::TextureFormat::required_features::h5d814caf5b611efe:
->  0x1013c70dc <+20>: ldrb   w11, [x9, x8]
    0x1013c70e0 <+24>: add    x10, x10, x11, lsl #2
    0x1013c70e4 <+28>: br     x10
    0x1013c70e8 <+32>: ret
Target 0: (bevyregress) stopped.
(lldb) bt
* thread #1, name = 'main', queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x4b8c64f1019f8cf8)
  * frame #0: 0x00000001013c70dc bevyregress`wgpu_types::TextureFormat::required_features::h5d814caf5b611efe + 20
    frame #1: 0x00000001013cdb70 bevyregress`wgpu_core::device::resource::Device$LT$A$GT$::create_texture::hbdd987e61dd6d7c5 + 212
    frame #2: 0x0000000101393f94 bevyregress`wgpu_core::device::global::_$LT$impl$u20$wgpu_core..global..Global$LT$G$GT$$GT$::device_create_texture::h73a7bf7e89af0308 + 348
    frame #3: 0x00000001013f8684 bevyregress`_$LT$wgpu..backend..direct..Context$u20$as$u20$wgpu..context..Context$GT$::device_create_texture::h81dbb4a560f1aead + 316
    frame #4: 0x00000001014013e4 bevyregress`_$LT$T$u20$as$u20$wgpu..context..DynContext$GT$::device_create_texture::h7fac5a1728c0e87e + 60
    frame #5: 0x00000001014a63a8 bevyregress`_$LT$wgpu..Device$u20$as$u20$wgpu..util..device..DeviceExt$GT$::create_texture_with_data::h91ea9c74edbd4d11 + 168
    frame #6: 0x0000000100866444 bevyregress`bevy_render::texture::fallback_image::fallback_image_new::h25cda84b669fe296 (.llvm.2045819193268999333) + 548
    frame #7: 0x0000000100866be4 bevyregress`_$LT$bevy_render..texture..fallback_image..FallbackImage$u20$as$u20$bevy_ecs..world..FromWorld$GT$::from_world::ha952e13675380366 + 1032
    frame #8: 0x0000000100981d44 bevyregress`_$LT$bevy_render..texture..ImagePlugin$u20$as$u20$bevy_app..plugin..Plugin$GT$::finish::h832044bf73ce2c5e + 1204
    frame #9: 0x00000001000e3490 bevyregress`bevy_app::app::App::finish::hba9fb047755ddcd5 + 104
    frame #10: 0x00000001000e31ec bevyregress`bevy_app::app::App::run::h39b5f7f41472f60b + 156
    frame #11: 0x0000000100fa00c8 bevyregress`bevyregress::main::hdc8715a225225966 + 128
    frame #12: 0x0000000100fa147c bevyregress`std::sys_common::backtrace::__rust_begin_short_backtrace::h55cb9fdb91c84ca5 + 12
    frame #13: 0x0000000100f9d89c bevyregress`std::rt::lang_start::_$u7b$$u7b$closure$u7d$$u7d$::h9fc2206eb4cf6bb0 + 16
    frame #14: 0x00000001012e4fd0 bevyregress`std::rt::lang_start_internal::hd092a83c379773e8 + 636
    frame #15: 0x0000000100f9d880 bevyregress`std::rt::lang_start::hf53aafa686c05f03 + 44
    frame #16: 0x0000000100fa0154 bevyregress`main + 32
    frame #17: 0x000000018fb410e0 dyld`start + 2360

```

Seems to be an invalid memory access in `wgpu_types::TextureFormat::required_features`?

**Comment by SkiFire13** [comment]
> That makes me wonder if Bevy is doing this thing that Embassy is (was?) doing: #117047

Bevy does something similar for its `apply_deferred` system, but it doesn't rely on function pointer equality, instead it erases the function item into a trait object and uses its `TypeId` for checking, which shouldn't suffer from the same problem.

**Comment by Patryk27** [comment]
It looks like disabling cross-crate inlining "solves" the issue, i.e.:
```
RUSTFLAGS='-Zcross-crate-inline-threshold=0' cargo run --release
```

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23117902.20Miscompilation.20of.20Bevy.20apps.20resulting.20in.20segfault.20o.E2.80.A6)).

@rustbot label -I-prioritize +P-high

**Comment by lqd** [comment]
Console-only reproducer:

```rust
fn main() {
    let mut group = bevy::app::PluginGroupBuilder::start::<bevy::app::NoopPluginGroup>();
    group = group
        .add(bevy::core::TaskPoolPlugin::default())
        .add(bevy::asset::AssetPlugin::default())
        .add(bevy::window::WindowPlugin::default()) // not required for segfault on release but running in debug
        .add(bevy::time::TimePlugin) // not required for segfault on release but running in debug
        .add(bevy::core::FrameCountPlugin) // not required for segfault on release but running in debug
        .add(bevy::render::RenderPlugin::default())
        .add(bevy::render::texture::ImagePlugin::default());

    bevy::app::App::new().add_plugins(group).run();
}
```

**Comment by artokun** [comment]
I can 1+ this, I was just doing the bevy tutorial and got to the part where I add DefaultPlugins leading to a seg fault. Once I added `RUSTFLAGS='-Zcross-crate-inline-threshold=0' cargo run` it recompiled and ran as expected.

**Comment by lqd** [comment]
After reducing `bevy`'s 150KLOC and 200+ dependencies: the miscompilation is not in `bevy`, but indeed likely in `wgpu 0.17`.

Here's a more minimal reproducer:

```toml
[dependencies]
wgpu = { version = "0.17.1"}
futures-lite = "1.12"
```

```rust
use wgpu::util::DeviceExt;

fn main() {
    let async_renderer = async move {
        let instance = wgpu::Instance::default();
        let adapter = instance
            .request_adapter(&Default::default())
            .await
            .expect("couldn't get instance");

        let (device, queue) = adapter
            .request_device(&Default::default(), None)
            .await
            .expect("couldn't get device and queue");

        let texture_descriptor = wgpu::TextureDescriptor {
            size: Default::default(),
            format: wgpu::TextureFormat::Rgba8UnormSrgb,
            dimension: wgpu::TextureDimension::D2,
            label: None,
            mip_level_count: 1,
            sample_count: 1,
            usage: wgpu::TextureUsages::TEXTURE_BINDING | wgpu::TextureUsages::COPY_DST,
            view_formats: &[],
        };

        device.create_texture_with_data(&queue, &texture_descriptor, &[255; 10]);
        panic!("non-deterministic failure should have happened above");
    };

    // Apparently, the executor needs to be multithreaded for this to reproduce.
    futures_lite::future::block_on(async_renderer);
}
```

There's some non-determinism involved, and likely from parallelism (and could depend on the hardware involved I'm not sure). The above can:
- work
- fail with a segfault
- fail with a bus error
- fail with a wgpu validation error about the texture format being unusable on the device
- fail with an abort during some metal deallocation,`-[_MTLCommandEncoder dealloc]:131: failed assertion `Command encoder released without endEncoding'`
- fail with an abort during a free, `regression-117902(58954,0x104ee8580) malloc: *** error for object 0x16b686690: pointer being freed was not allocated`

For `bevy`, updating to `wgpu 0.18` *seems* to work around the issue. At least, that was enough for me on the reproducer above.

If someone wants to help, bisecting `wgpu` from 0.17.x to 0.18 should at least give an indication of what happened to change this behavior here (could be UB and/or missing synchronization, but just looking at the open issues there we can see mentions of data races and missing synchronization when using metal). That may help see why cross-crate inlining was involved here.

**Comment by saethlin** [comment]
If it is a data race, -Zsanitizer=thread might be informative.

**Comment by saethlin** [comment]
(I would be trying these myself but I just have a phone for a bit)

Can you bisect along values of -Ccodegen-units and -Zcross-crate-inline-threshold? I would bisect between 1 and 1024 for both. If you can find a cross crate threshold that's required it wouldn't be difficult to stick some prints into it compiler to see what DefId is changing codegen mode.

**Comment by inodentry** [comment]
@cwfitzgerald @kvark (any other wgpu maintainers?)

Do you have any knowledge/ideas that might help with figuring out why rustc might be miscompiling wgpu?

**Comment by cwfitzgerald** [comment]
While we do have various races in places, that code is should be rather straightforward - no immediate ideas why that would be miscompiled.

**Comment by lqd** [comment]
This may have been fixed/worked around by https://github.com/gfx-rs/wgpu/pull/4008.

Here are a few [asan and tsan runs](https://gist.github.com/lqd/c5ebebedb75a82e4e2bd37aae01abc4a), but weren't enlightening to me. Maybe Ben will see something.

**Comment by saethlin** [comment]
Trying to understand the behavior of LLVM given such a large program which is known to execute UB (data races and now null pointer dereferences) seems way too difficult to me. If I was working this (I'm not because it's holiday time) I'd fix those bugs then reasses.

**Comment by apiraino** [comment]
Raising the priority to signal that we want to track this and ensure it does not land on stable (wherever the root cause)

@rustbot labl -P-high +P-critical

**Comment by saethlin** [comment]
The cross-crate-inlining change I made essentially infers the presence of the `#[inline]` attribute where it is deduced to be a reasonable addition.

I followed my own suggestion above to bisect along values of `-Zcross-crate-inline-threshold` and deduced that the magic value is 21. Then I stuck some code into the pass to print every candidate with a value of 21. Only a handful were printed in the build. I hacked up the contents of my `~/.cargo/registry/src` to add `#[inline(never)]` to each of them in turn.

The magic function is this: https://github.com/gfx-rs/wgpu/blob/21098cdaceee49c22a7e3a9648ce019b4e74beec/wgpu-types/src/lib.rs#L2661

If I add `#[inline]` to that function, @lqd's reproducer above will segfault on stable.

**Comment by saethlin** [comment]
`required_features` leads me to `describe_format_features`, which if I put `#[inline(never)]` on prevents the segfaults. But above that is `create_texture` which is huge and does a lot of things, and adding `#[inline(never)]` to that function has no effect. So I think the unexpected codegen happens while optimizing `wgpu_core::Device::create_texture`, which at least I suppose narrows things down from the whole crate graph.

**Comment by inodentry** [comment]
Oh ... great ... so this is a miscompilation bug that also exists in stable, which could have been triggered if some specific code was inlined. Now rustc just inlines it automatically. So it isn't a codegen regression, but an old/preexisting bug exposed by new rustc enabling more aggressive optimizations.

Damn. I'm so glad I reported this. Never could have imagined that it would go so deep...

**Comment by saethlin** [comment]
I am not yet convinced that this is a miscompilation as opposed to UB that is now exploited.

**Comment by lqd** [comment]
> `required_features` [...] `wgpu_core::Device::create_texture`

These 2 functions arethe the closest to the segfault in the backtrace, so the problem surely lies around there. We could likely extract them both and hopefully repro outside of wgpu.

**Comment by saethlin** [comment]
I have not managed to separate the strangely-behaving code out from wgpu, but I have learned some things.

`required_features` is very strange. If I stick `#[inline(never)]` on it, LLVM generates this function signature:
```
define noundef i64 @_ZN10wgpu_types13TextureFormat17required_features17hd1a6387d4186056bE(ptr noalias nocapture noundef readonly align 4 dereferenceable(12) %self) unnamed_addr #6 !dbg !2239 {
```
Which is exactly what you'd expect. But if I remove the attribute and leave cross-crate-inlining enabled, or add `#[inline]` we get this signature:
```
define internal fastcc noundef i64 @_ZN10wgpu_types13TextureFormat17required_features17hd1a6387d4186056bE(i32 %self.0.val, i32 %self.8.val) unnamed_addr #1 !dbg !24733 {
```
So LLVM has done some kind of cross-function optimization here. The IR for the function bodies looks exactly the same to me, except for the fact that one does two gepis and the other just uses its arguments.

And of course they are lowered to slightly different assembly. Here is the `#[inline(never)]` one:
```
100225494: aa0003e8     mov     x8, x0
100225498: b9400009     ldr     w9, [x0]
10022549c: d2800000     mov     x0, #0
1002254a0: 9000058a     adrp    x10, 0x1002d5000
1002254a4: 911b414a     add     x10, x10, #1744
1002254a8: 1000008b     adr     x11, #16
1002254ac: 3869694c     ldrb    w12, [x10, x9]
1002254b0: 8b0c096b     add     x11, x11, x12, lsl #2
1002254b4: d61f0160     br      x11
```

And the `#[inline]` one:
```
100053860: aa0003e8     mov     x8, x0
100053864: d2800000     mov     x0, #0
100053868: 90001289     adrp    x9, 0x1002a3000
10005386c: 91032529     add     x9, x9, #201
100053870: 1000008a     adr     x10, #16
100053874: 3868692b     ldrb    w11, [x9, x8]
100053878: 8b0b094a     add     x10, x10, x11, lsl #2
10005387c: d61f0140     br      x10
```
Directly after both of these is a small jump table. But I think what's down there is irrelevant, because in the `#[inline]` one, when we `br` on it, `x10` contains an address that is comfortably outside this function.

So I think whatever is going on here relies on `required_features` getting this SRoA codegen, but also not be inlined.

---

lldb has been supremely anti-helpful. It says breakpoints are set then blows past them (realizing this was happening took me a long time), and if I do too much instruction stepping it segfaults.

**Comment by Nilstrieb** [comment]
I think what's happening with the signature is argument promotion, where it replaces an indirect argument with a direct one.

**Comment by lqd** [comment]
I've removed 99% of `wgpu` in https://github.com/lqd/repro-117902, but it still currently clocks in at 1000 lines or so.

> I have not managed to separate the strangely-behaving code out from wgpu

It's indeed way harder than one would hope. It's quite fiddly, maybe there's a magic value of the inlining threshold for different shapes, combined with various inline never/always attributes, but I didn't manage to merge more of the crates together; there are still 2 `wgpu-*` crates in addition to the main binary.

There are still a bunch of backends/API/etc wgpu abstractions in the repro so it can still be reduced more, but I'm not sure I'll have much more time to dedicate to this. At the very least, there are still two dependencies out of three that shouldn't be hard to remove (`thiserror` and `bitflags`) .
But there are limits/roadblocks to how much we can minimize without having thresholds/inline attributes that make this reproduce all the time, as some of these sizes impact inlining decisions.

I still don't know if this is a miscompile or UB:
- if it's the latter, it's maybe not in `wgpu`: there's no unsafe code in the repro per se, but there's surely some in `parking_lot`.
- it's possible that `parking_lot` on mac exhibits an issue though. I *think* I removed all the osx/metal specific code, it could be cool to test on aarch64-linux (maybe on the dev-desktops) and on an intel mac, to see what happens there. 
- switching to the std mutex and rwlock also made the segfault disappear but it's not clear whether it could just be some different behavior or different codegen that would stop triggering the issue.
- the number of variants in the `TextureFormat` enum seem to matter. There needs to be quite a lot or the problem disappears.

**Comment by saethlin** [comment]
My latest progress brings the line count down to 221 lines, and there is precious little logic left, it's mostly bitflags and enum definitions. It's in a PR to the above repo. It got way more consistent when I found the path to remove the locks.

I'm sure the remaining bits of `wgpu_types` could be cleaned up. And I didn't try removing any of the crate boundaries, though I suspect those are required.

At this point I'm quite sure this is a miscompile. There's no unsafe code left in the repo, and Miri doesn't complain.

I've tried running this on aarch64-unknown-linux-gnu and x86_64-unknown-linux-gnu, and neither segfault. I doubt this is an OS issue, perhaps there is something different between the default aarch64-apple-darwin target and the aarch64-unknown-linux-gnu target that would be enlightening.

**Comment by saethlin** [comment]
Also somehow the MIR inliner is still required.
```
RUSTFLAGS="-Zmir-opt-level=0 -Zinline-mir -Zinline-mir-hint-threshold=12 -Zinline-mir-threshold=0" cargo r --release
```
Though it's not clear if that's because the bug is in the MIR inliner or if only post-inlining do we generate the exact LLVM IR pattern that gets miscompiled.

**Comment by lqd** [comment]
The repro is now at around 150 lines, with 1 bin and 1 lib. It's minimal enough e.g. for a test or LLVM bug report.

> Also somehow the MIR inliner is still required.

@saethlin Doesn't `-Zinline-mir=no` fully turn off the inliner? Because the segfault reproduces for me even with this flag.

In any case, the LLVM optimization pass that starts introducing the segfault is `Machine code sinking on function (core::ops::function::FnOnce::call_once{{vtable.shim}})`, `RUSTFLAGS="-Cllvm-args=-opt-bisect-limit=1255"` to reproduce this on the latest revision.

**Comment by ActuallyHappening** [comment]
It's nice to know I'm not the only one experiencing random segfaults in my bevy apps on `nightly`, and that it is being actively explored!
I have no `rustc` dev experience to offer, but I did notice another compiler bug ([original](https://github.com/rust-lang/rust/issues/118030)) [here](https://github.com/rust-lang/rust/issues/118025), but I'm pretty sure they are unrelated.

**Comment by DianQK** [comment]
Upstream issue: https://github.com/llvm/llvm-project/issues/74680.
I tried to submit a fix as well.

Base on https://github.com/lqd/repro-117902, I got
```rust
fn main() {
    call_from_main();
}

#[inline(never)]
#[no_mangle]
fn call_from_main() {
    let desc = TextureDescriptor(TextureFormat::Bc6hRgbUfloat);
    device_create_texture(desc);
}

#[inline(never)]
#[no_mangle]
pub fn device_create_texture(desc: TextureDescriptor) {
    let format = desc.0;
    format.required_features().used();
    format.used();
}

enum Feature {
    A,
    B,
}

impl Feature {
    #[inline(never)]
    fn used(self) {
        core::hint::black_box(self);
    }
}

#[repr(C)]
pub enum AstcBlock {
    B12x12,
}
pub enum AstcChannel {
    Unorm,
    Hdr,
}
#[repr(C)]
pub enum TextureFormat {
    Rgba8UnormSrgb,
    Bc6hRgbUfloat,
    Key,
    Astc {
        channel: AstcChannel,
        block: AstcBlock,
    },
}
impl TextureFormat {
    #[inline(never)]
    fn used(&self) {
        core::hint::black_box(self);
    }
    #[inline(never)]
    // #[no_mangle]
    fn required_features(&self) -> Feature {
        match self {
            Self::Rgba8UnormSrgb => Feature::A,
            Self::Bc6hRgbUfloat => Feature::B,
            Self::Key => Feature::B,
            Self::Astc { channel, .. } => match channel {
                AstcChannel::Hdr => Feature::A,
                AstcChannel::Unorm => Feature::B,
            },
        }
    }
}
pub struct TextureDescriptor(pub TextureFormat);
```

**Comment by DianQK** [comment]
Hmm, I can only use nightly to reproduce, and I can't use local stage1/stage2 to reproduce. This could be a difference in build configuration. So I think this minimal reproduction is unstable. If we want to use this as a test case, I would expect a more stable test case.

I added a more detailed explanation of this issue at https://github.com/llvm/llvm-project/pull/74682#issuecomment-1846339128. We should be able to create a must-be-dirty stack to reproduce.

**Comment by lqd** [comment]
Reopening to track beta backport

**Comment by lqd** [comment]
This can now be closed: this issue was fixed on nightly by the cherry-picked LLVM fix in #118936, and fixed on beta by backporting that PR in #118994, so it thankfully won't reach stable in a couple weeks. Thanks in particular to @saethlin and @DianQK.
